### DISCUSSION TITLE
Have you ever faced crashes while sending nested actions synchronously?

### DISCUSSION THREAD

**Discussion by maximkrouk** [pointfreeco/swift-composable-architecture#449] [Q&A] [answered]
I use https://github.com/trading-point/reactiveswift-composable-architecture with UIKit and I had to wrap a lot of `viewStore.send` calls into `DispatchQueue.main.async` call, because `viewStore.send` caused crashes in `@MutableProperty var state` (== `@Published`) in reducer due to the deadlock in the private `UnfairLock`. I guess it's just a ReactiveSwift's implementation that causes these problems, but I'm not sure, so I decided to make a DisscussionThread/Poll on that topic.

For now I'm waiting for `OpenCombine` to add `Publishers.Merge` to make a port of TCA asap, but I'd like to improve my hopes on stability by this thread 😅

**Comment by mluisbrown** [comment, marked as answer]
I was discussing this with a colleague of mine (@p4checo) the other day and he experimented doing the same thing using `Combine` and it apparently does not crash. So somehow Combine allows changing the `value` of a `CurrentValueSubject` without locking 🤯 

I looked at the implementation of `CurrentValueSubject` in [OpenCombine](https://github.com/OpenCombine/OpenCombine/blob/master/Sources/OpenCombine/CurrentValueSubject.swift) and it also uses an `UnfairLock` just like `ReactiveSwift`, so I'm not quite sure how `Combine` manages this without locks.

**Reply by mluisbrown** [reply]
Actually I did some experiments with OpenCombine, and it doesn't have the problem, despite using locks. In it's `CurrentValueSubject` it release the lock before publishing values to its subscribers. The lock is only around the modification of the value itself. I guess that is how Combine does it too.

I'm going to see if I can do something similar in RAS-TCA, by using something custom instead of `MutableProperty` for the state.

**Reply by maximkrouk** [reply]
Actually that makes sense, glad u figured it 😍

**Reply by mluisbrown** [reply]
FYI: I have a [WIP PR](https://github.com/trading-point/reactiveswift-composable-architecture/pull/41) of RAS-TCA where I'm replicating this behaviour of `CurrentValueSubject`, in that the sending of updates to observers is _outside_ the scope of the lock.

I haven't tried it yet to see if it solves the problem. If you have time @maximkrouk it would be great if you could try it in your project, as you seem to be having this issue a lot.
